<a href="https://colab.research.google.com/github/Nishpowale/Book_recommender_System_using_Clustering/blob/main/Book_recommender_System_using_Clustering.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Collaborative filtering is also known as social filtering. Collaborative filtering uses algorithms to filter data from user reviews to make personalized recommendations for users with similar preferences. Collaborative filtering is also used to select content and advertising for individuals on social media.

In [69]:
import pandas as pd#for handling the data set 
import numpy as np # for mathematical calculations
import seaborn as sns#for Visualization 
import matplotlib.pyplot as plt# For Visualization 
import sklearn # for model evaluation 
%matplotlib inline

##Importing the data sets
----




In [70]:
Books_df=pd.read_csv('BX-Books.csv',sep=';',error_bad_lines=False,encoding='latin-1')

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2882: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version.


  exec(code_obj, self.user_global_ns, self.user_ns)
b'Skipping line 6452: expected 8 fields, saw 9\nSkipping line 43667: expected 8 fields, saw 10\nSkipping line 51751: expected 8 fields, saw 9\n'
b'Skipping line 92038: expected 8 fields, saw 9\nSkipping line 104319: expected 8 fields, saw 9\nSkipping line 121768: expected 8 fields, saw 9\n'
b'Skipping line 144058: expected 8 fields, saw 9\nSkipping line 150789: expected 8 fields, saw 9\nSkipping line 157128: expected 8 fields, saw 9\nSkipping line 180189: expected 8 fields, saw 9\nSkipping line 185738: expected 8 fields, saw 9\n'
b'Skipping line 209388: expected 8 fields, saw 9\nSkipping line 220626: expected 8 fields, saw 9\nSkipping line 227933: expected 8 fields, saw 11\nSkipping line 228957: expected 8 fields, saw 10\nSkipping line 245933: expected 

The reason for using separation ';' is that on looking up the data set we can see the values are separated using a semi colon 

Certain lines have been skipped because the number of expected fields are 8 but got 9 

In [71]:
Books_df.head()

ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            Book-Author Year-Of-Publication                   Publisher  \
0    Mark P. O. Morford                2002     Oxford University Press   
1  Richard Bruce Wright                2001       HarperFlamingo Canada   
2          Carlo D'Este                1991             HarperPerennial   
3      Gina Bari Kolata                1999        Farrar Straus Giroux   
4       E. J. W. Barber                1999  W. W. Norton &amp; Company   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

Here the ISBN number is the International Standard Book Number (ISBN)  which is a international standard identification number for uniquely identifying books.

In [72]:
Books_df.tail()

ISBN                                         Book-Title  \
271355  0440400988                         There's a Bat in Bunk Five   
271356  0525447644                            From One to One Hundred   
271357  006008667X  Lily Dale : The True Story of the Town that Ta...   
271358  0192126040                        Republic (World's Classics)   
271359  0767409752  A Guided Tour of Rene Descartes' Meditations o...   

                Book-Author Year-Of-Publication  \
271355       Paula Danziger                1988   
271356           Teri Sloat                1991   
271357     Christine Wicker                2004   
271358                Plato                1996   
271359  Christopher  Biffle                2000   

                                               Publisher  \
271355                   Random House Childrens Pub (Mm)   
271356                                      Dutton Books   
271357                                HarperSanFrancisco   
271358                           Oxford University Press   
271359  McGraw-Hill Humanities/Social Sciences/Languages   

                                              Image-URL-S  \
271355  http://images.amazon.com/images/P/0440400988.0...   
271356  http://images.amazon.com/images/P/0525447644.0...   
271357  http://images.amazon.com/images/P/006008667X.0...   
271358  http://images.amazon.com/images/P/0192126040.0...   
271359  http://images.amazon.com/images/P/0767409752.0...   

                                              Image-URL-M  \
271355  http://images.amazon.com/images/P/0440400988.0...   
271356  http://images.amazon.com/images/P/0525447644.0...   
271357  http://images.amazon.com/images/P/006008667X.0...   
271358  http://images.amazon.com/images/P/0192126040.0...   
271359  http://images.amazon.com/images/P/0767409752.0...   

                                              Image-URL-L  
271355  http://images.amazon.com/images/P/0440400988.0...  
271356  http://images.amazon.com/images/P/0525447644.0...  
271357  http://images.amazon.com/images/P/006008667X.0...  
271358  http://images.amazon.com/images/P/0192126040.0...  
271359  http://images.amazon.com/images/P/0767409752.0...

In [73]:
Books_df.columns

Index(['ISBN', 'Book-Title', 'Book-Author', 'Year-Of-Publication', 'Publisher',
       'Image-URL-S', 'Image-URL-M', 'Image-URL-L'],
      dtype='object')

In [74]:
Books_df=Books_df.drop(['Image-URL-S', 'Image-URL-M', 'Image-URL-L'], axis=1)


In [75]:
Books_df.head(1)

ISBN           Book-Title         Book-Author Year-Of-Publication  \
0  0195153448  Classical Mythology  Mark P. O. Morford                2002   

                 Publisher  
0  Oxford University Press

In [76]:
Books_df.rename(columns={'Book-Title':'title','Book-Author':'author','Year-Of-Publication':'year'},inplace=True)

In [77]:
Books_df.head(1)

ISBN                title              author  year  \
0  0195153448  Classical Mythology  Mark P. O. Morford  2002   

                 Publisher  
0  Oxford University Press

---------
Loading the Users data set 


In [78]:
Users=pd.read_csv('BX-Users.csv',sep=';',error_bad_lines=False,encoding='latin-1')

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2882: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version.


  exec(code_obj, self.user_global_ns, self.user_ns)


In [79]:
Users.shape

(278858, 3)

In [80]:
Users.rename(columns={'User-ID':'user_id','Location':'location','Age':'age'})

user_id                            location   age
0             1                  nyc, new york, usa   NaN
1             2           stockton, california, usa  18.0
2             3     moscow, yukon territory, russia   NaN
3             4           porto, v.n.gaia, portugal  17.0
4             5  farnborough, hants, united kingdom   NaN
...         ...                                 ...   ...
278853   278854               portland, oregon, usa   NaN
278854   278855  tacoma, washington, united kingdom  50.0
278855   278856           brampton, ontario, canada   NaN
278856   278857           knoxville, tennessee, usa   NaN
278857   278858                dublin, n/a, ireland   NaN

[278858 rows x 3 columns]

----
Loading the Ratings Data Set 

In [81]:
Ratings=pd.read_csv('BX-Book-Ratings.csv',sep=';',error_bad_lines=False,encoding='latin-1')

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2882: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version.


  exec(code_obj, self.user_global_ns, self.user_ns)


In [82]:
Ratings.head(2)

User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5

In [83]:
Ratings.rename(columns={'Book-Rating':'rating'})

User-ID         ISBN  rating
0         276725   034545104X       0
1         276726   0155061224       5
2         276727   0446520802       0
3         276729   052165615X       3
4         276729   0521795028       6
...          ...          ...     ...
1149775   276704   1563526298       9
1149776   276706   0679447156       0
1149777   276709   0515107662      10
1149778   276721   0590442449      10
1149779   276723  05162443314       8

[1149780 rows x 3 columns]

In [84]:
Books_df.shape


(271360, 5)

In [85]:
Users.shape

(278858, 3)

In [86]:
Ratings.shape

(1149780, 3)

-----
Extracting the Data of Only the useful Users 

In [87]:
Ratings['User-ID'].value_counts()

11676     13602
198711     7550
153662     6109
98391      5891
35859      5850
          ...  
116180        1
116166        1
116154        1
116137        1
276723        1
Name: User-ID, Length: 105283, dtype: int64

We use value_counts to identify the number of ratings made by each user .
From the result , we can identify the users who have a high number of given ratings and thus we can use these particular users for recommeding as they will be more appropriate for book recommedations, 

In [88]:
Ratings['User-ID'].value_counts().shape

(105283,)

Out of the total number of users(, only 105283 have actually given a rating 





In [89]:
x=Ratings['User-ID'].value_counts()>200 

In [90]:
x[x]
#To take only those values which are true 

11676     True
198711    True
153662    True
98391     True
35859     True
          ... 
274808    True
28634     True
59727     True
268622    True
188951    True
Name: User-ID, Length: 899, dtype: bool

In [91]:
valid_users=x[x].index

In [92]:
valid_users

Int64Index([ 11676, 198711, 153662,  98391,  35859, 212898, 278418,  76352,
            110973, 235105,
            ...
            260183,  73681,  44296, 155916,   9856, 274808,  28634,  59727,
            268622, 188951],
           dtype='int64', length=899)

Now this extracted data of the useful users is needed to be added to the ratings data set .

In [93]:
Ratings=Ratings[Ratings['User-ID'].isin(valid_users)]

The .isin function return only those values which are present in the Ratings Column ' User-ID' and valid_users 

In [94]:
Ratings.shape


(526356, 3)

Our number of rows in Ratings has reduced down to 526356

----
Merging the ratings dataframe with the books data set 

In [95]:
ratings_with_books=Ratings.merge(Books_df,on='ISBN')
ratings_with_books.head(1)

User-ID        ISBN  Book-Rating  \
0   277427  002542730X           10   

                                               title             author  year  \
0  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner  1994   

                   Publisher  
0  John Wiley &amp; Sons Inc

Around 40,000 numbers have been reduced. Certain ISBN numbers must have no information about the book , hence have been discarded.

In [96]:
number_rating= ratings_with_books.groupby('title')['Book-Rating'].count().reset_index()

In [97]:
number_rating.rename(columns={'Book-Rating':'number of ratings'},inplace=True)
number_rating 

title  number of ratings
0        A Light in the Storm: The Civil War Diary of ...                  2
1                                   Always Have Popsicles                  1
2                    Apple Magic (The Collector's series)                  1
3        Beyond IBM: Leadership Marketing and Finance ...                  1
4        Clifford Visita El Hospital (Clifford El Gran...                  1
...                                                   ...                ...
160264  Ã?Â?ber die Pflicht zum Ungehorsam gegen den S...                  3
160265                                      Ã?Â?lpiraten.                  1
160266                     Ã?Â?rger mit Produkt X. Roman.                  1
160267                              Ã?Â?stlich der Berge.                  1
160268                                  Ã?Â?thique en toc                  1

[160269 rows x 2 columns]

In [98]:
final_rating=ratings_with_books.merge(number_rating, on ='title')
final_rating

User-ID        ISBN  Book-Rating  \
0        277427  002542730X           10   
1          3363  002542730X            0   
2         11676  002542730X            6   
3         12538  002542730X           10   
4         13552  002542730X            0   
...         ...         ...          ...   
487666   275970  1892145022            0   
487667   275970  1931868123            0   
487668   275970  3411086211           10   
487669   275970  3829021860            0   
487670   275970  4770019572            0   

                                                    title  \
0       Politically Correct Bedtime Stories: Modern Ta...   
1       Politically Correct Bedtime Stories: Modern Ta...   
2       Politically Correct Bedtime Stories: Modern Ta...   
3       Politically Correct Bedtime Stories: Modern Ta...   
4       Politically Correct Bedtime Stories: Modern Ta...   
...                                                   ...   
487666                                   Here Is New York   
487667  There's a Porcupine in My Outhouse: Misadventu...   
487668                                         Die Biene.   
487669                                     The Penis Book   
487670                                            Musashi   

                               author  year  \
0                   James Finn Garner  1994   
1                   James Finn Garner  1994   
2                   James Finn Garner  1994   
3                   James Finn Garner  1994   
4                   James Finn Garner  1994   
...                               ...   ...   
487666                    E. B. White  1999   
487667                   Mike Tougias  2002   
487668  Sybil Gr�¤fin Sch�¶nfeldt  1993   
487669                   Joseph Cohen  1999   
487670                 Eiji Yoshikawa  1995   

                                   Publisher  number of ratings  
0                  John Wiley &amp; Sons Inc                 82  
1                  John Wiley &amp; Sons Inc                 82  
2                  John Wiley &amp; Sons Inc                 82  
3                  John Wiley &amp; Sons Inc                 82  
4                  John Wiley &amp; Sons Inc                 82  
...                                      ...                ...  
487666                       Little Bookroom                  1  
487667                    Capital Books (VA)                  1  
487668  Bibliographisches Institut, Mannheim                  1  
487669                              Konemann                  1  
487670          Kodansha International (JPN)                  1  

[487671 rows x 8 columns]

In [99]:
final_rating.shape

(487671, 8)

In [100]:
final_rating=final_rating[final_rating['number of ratings']>=50]
#final_rating[final_rating] this means to take those values which are true adn then we put it bak in the df 
final_rating.shape

(61853, 8)

We have taken only those Users who have given more than 200 ratings and those books which have more than 50 ratings

In [102]:
#Now we drop any duplicate users and titles which are present 
final_rating.drop_duplicates(['User-ID','title'],inplace=True)

/usr/local/lib/python3.7/dist-packages/pandas/util/_decorators.py:311: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return func(*args, **kwargs)


In [103]:
final_rating.shape

(59850, 8)

----
Making a pivot table 

A pivot table is a similar operation that is commonly seen in spreadsheets and other programs that operate on tabular data. The pivot table takes simple column-wise data as input, and groups the entries into a two-dimensional table that provides a multidimensional summarization of the data.

In [106]:
book_pivot=final_rating.pivot_table(columns='User-ID',index='title',values='Book-Rating')
book_pivot

User-ID                                             254     2276    2766    \
title                                                                        
1984                                                   9.0     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN    10.0     NaN   
4 Blondes                                              NaN     NaN     NaN   
84 Charing Cross Road                                  NaN     NaN     NaN   
...                                                    ...     ...     ...   
Year of Wonders                                        NaN     NaN     NaN   
You Belong To Me                                       NaN     NaN     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     NaN     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

User-ID                                             2977    3363    3757    \
title                                                                        
1984                                                   NaN     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN     NaN     NaN   
4 Blondes                                              NaN     NaN     NaN   
84 Charing Cross Road                                  NaN     NaN     NaN   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     NaN     NaN   
You Belong To Me                                       NaN     NaN     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     0.0     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

User-ID                                             4017    4385    6242    \
title                                                                        
1984                                                   NaN     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN     NaN     NaN   
4 Blondes                                              NaN     NaN     NaN   
84 Charing Cross Road                                  NaN     NaN     NaN   
...                                                    ...     ...     ...   
Year of Wonders                                        NaN     NaN     7.0   
You Belong To Me                                       NaN     NaN     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     NaN     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

User-ID                                             6251    ...  274004  \
title                                                       ...           
1984                                                   NaN  ...     NaN   
1st to Die: A Novel                                    NaN  ...     NaN   
2nd Chance                                             NaN  ...     NaN   
4 Blondes                                              0.0  ...     NaN   
84 Charing Cross Road                                  NaN  ...     NaN   
...                                                    ...  ...     ...   
Year of Wonders                                        NaN  ...     NaN   
You Belong To Me                                       NaN  ...     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     0.0  ...     NaN   
Zoya                                                   NaN  ...     NaN   
\O\" Is for Outlaw"                                    

We have 742 Books and 888 users out of the entire data set now.
Now to remove the empyty values we replace them with 0 .

In [108]:
book_pivot.fillna(0, inplace=True)
book_pivot.head(1)

User-ID  254     2276    2766    2977    3363    3757    4017    4385    \
title                                                                     
1984        9.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   

User-ID  6242    6251    ...  274004  274061  274301  274308  274808  275970  \
title                    ...                                                   
1984        0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   

User-ID  277427  277478  277639  278418  
title                                    
1984        0.0     0.0     0.0     0.0  

[1 rows x 888 columns]

----
Model Development 


Nearest neighbor search (NNS), as a form of proximity search, is the optimization problem of finding the point in a given set that is closest (or most similar) to a given point.(Clustering Algorithm - It finds out the distances and puts the nearest ones together to form a cluster )

In [110]:
from scipy.sparse import csr_matrix
book_sparse=csr_matrix(book_pivot)

Python's SciPy provides tools for creating sparse matrices using multiple data structures, as well as tools for converting a dense matrix to a sparse matrix. The sparse matrix representation outputs the row-column tuple where the matrix contains non-zero values along with those values.

In [113]:
type(book_sparse)

scipy.sparse.csr.csr_matrix

What it does is basically it dosent consider the zero values and only considers non zero values , as finding distances of zero values will add addiotional computational time .

In [116]:
from sklearn.neighbors import NearestNeighbors 
model=NearestNeighbors(algorithm ='brute')
#brute means taking all iterations 

In [118]:
model.fit(book_sparse)
#Model has been trained 

NearestNeighbors(algorithm='brute')

In [120]:
distances, suggestions = model.kneighbors(book_pivot.iloc[237 , :].values.reshape(1,-1),n_neighbors=6)
#237 is the ISBN number of Harry Potter 

When it comes to selecting rows and columns of a pandas DataFrame, loc and iloc are two commonly used functions. Here is the subtle difference between the two functions: loc selects rows and columns with specific labels. iloc selects rows and columns at specific integer positions

In [121]:
distances

array([[ 0.        , 68.78953409, 69.5413546 , 72.64296249, 76.83098333,
        77.28518616]])

Here we get the distances of the closest 5 books 

In [122]:
suggestions

array([[237, 240, 238, 241, 184, 536]])

In [123]:
for i in range(len(suggestions)):
  print(book_pivot.index[suggestions[i]])

Index(['Harry Potter and the Chamber of Secrets (Book 2)',
       'Harry Potter and the Prisoner of Azkaban (Book 3)',
       'Harry Potter and the Goblet of Fire (Book 4)',
       'Harry Potter and the Sorcerer's Stone (Book 1)', 'Exclusive',
       'The Cradle Will Fall'],
      dtype='object', name='title')


In [124]:
book_pivot.index[240]

'Harry Potter and the Prisoner of Azkaban (Book 3)'

----
Creating a Fucntion where the user can input the name and have recommenddations 

In [129]:
#How to get the id of the books when user enters the name of the book
book_pivot.index=='Harry Potter and the Prisoner of Azkaban (Book 3)'
#This results in an array where every element shows flase except the one place where it is true 

array([False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False,

In [130]:
#Now to extract the true value we use the numpy function np.where which returns only the true value 
np.where(book_pivot.index=='Harry Potter and the Prisoner of Azkaban (Book 3)')

(array([240]),)

In [132]:
#To extract the Book id 240 we do 
np.where(book_pivot.index=='Harry Potter and the Prisoner of Azkaban (Book 3)')[0][0]

240

In [164]:
#How to get the id of the books when user enters the name of the book
def recommedbooks (book_name):
  book_id=np.where(book_pivot.index==book_name)[0][0]
  distances, suggestions = model.kneighbors(book_pivot.iloc[book_id , :].values.reshape(1,-1),n_neighbors=6)

  for i in range(len(suggestions)):
    #if i != 0 :#To no print the book entered 
    suggestions_list=book_pivot.index[suggestions[i]]
    print('The suggested recommendations for',book_name,'are as follows:')
    print()
    count=1
    for count,recom in enumerate(suggestions_list[1:],1):
      print(' ' ,count , recom)
      count+=1

----
Testing the model 

In [168]:
recommedbooks('The Da Vinci Code')

The suggested recommendations for The Da Vinci Code are as follows:

  1 Touching Evil
  2 The Blue Nowhere : A Novel
  3 Saving Faith
  4 Zoya
  5 Sea Glass: A Novel


In [181]:
recommedbooks('1984')

The suggested recommendations for 1984 are as follows:

  1 No Safe Place
  2 A Civil Action
  3 Foucault's Pendulum
  4 Long After Midnight
  5 Abduction


In [183]:
recommedbooks('Long After Midnight')

The suggested recommendations for Long After Midnight are as follows:

  1 No Safe Place
  2 Exclusive
  3 Last Man Standing
  4 Lake Wobegon days
  5 Pleading Guilty
